### ISSUE TITLE
[Bug] Shaky modeline

### ISSUE THREAD

**Issue by memeplex** [seagle0128/doom-modeline#478]
### Thank you for the bug report

- [X] I am using the latest version of `doom-mode` related packages.
- [X] I checked [FAQ](https://github.com/seagle0128/doom-modeline#faq).
- [X] You may also try reproduce the issue using clean environment and minimal configurations with
the command `emacs -Q`.

### Bug description

This started as a FR (that stands independently of this) to move the checker in order to avoid jerky distracting displacements in the modeline contents. See https://github.com/seagle0128/doom-modeline/issues/477 for further context and a screencast showing the issue.

But even after putting the checker at the left of the rhs, other segments are unnecessarily re-accommodating all the time.

### Steps to reproduce

1. Set the default modeline to have 'checker as the first item in the rhs
2. Open a python file
3. Enable flycheck
4. Introduce a number of errors
5. Delete some of them
6. Quickly repeat 3 and 4
 

### Expected behavior

Segments that haven't changed themselves nor segments to their right, stay quiet avoiding unnecessary distractions and unpleasant visual artifacts.

### OS

MacOS

### Emacs Version

28

### Emacs Configurations

Custom

**Comment by seagle0128** [comment]
How about disabling icons? `(setq doom-modeline-icon nil)`.

**Comment by memeplex** [comment]
It seems too radical a solution, a middle ground alternative is to allow disabling of this one:

```
('running     (doom-modeline-checker-icon "access_time" "⏱" "*" 'doom-modeline-debug))
```

This is constantly triggering transitions. I mean, every little modification of the buffer trigger this one.

OTOH there is something else going on here, because the transitions happen at the left of the other icons, this is different to #477. In #477 there is a good reason why the displacements happen, but I don't see one here.

**Comment by memeplex** [comment]
I've been investigating the more difficult problems with displacements when icons are enabled, for example the one described above:

* `A B` at the rhs, a change in `A` shouldn't produce displacements in `B`.

but also things like in the checker:

* `n <access_time-icon>` -> `n <block-icon>` produces a displacement just because of the icon change, even if `n` is constant.

All these issues stem from:

```elisp
               'display `((space
                           :align-to
                           (- (+ right right-fringe right-margin)
                              ,(* (let ((width (doom-modeline--font-width)))
                                    (or (and (= width 1) 1)
                                        (/ width (frame-char-width) 1.0)))
                                  (string-width
                                   (format-mode-line (cons "" rhs-forms))))))))
```

In particular `string-width` is very unreliable with special characters, the string `<space><block-icon><space>` and the string `<space><access_time-icon><space>` are measured as having widths 3 and 4 respectively, although both of them contain three characters and  have the same pixel width. The documentation of `string-width` warns about this fact:

> When calculating width of a multibyte character in STRING,
only the base leading-code is considered; the validity of
the following bytes is not checked.  Tabs in STRING are always
taken to occupy ‘tab-width’ columns.  The effect of faces and fonts
used for non-Latin and other unusual characters (such as emoji) is
ignored as well, as are display properties and invisible text.
For these reasons, the results are not generally reliable;
for accurate dimensions of text as it will be displayed,
use ‘window-text-pixel-size’ instead.

The problem is that `window-text-pixel-size` requires that the text to be measured is already in a live window, so it seems pretty useless in this context.

I don't think there is much to be done in this regard, someday we will have good svg icons that aren't a PITA to align, I guess.

So forget about the icon-enabled case. The text-only case is still worth tackling, I've described a very simple solution [in this comment](https://github.com/seagle0128/doom-modeline/issues/478#issuecomment-942695245).

**Comment by seagle0128** [comment]
Closed by 1d400aa.
